FEAT: [TESTMERGE] New Goal - Bluespace Rift Scanner

PhantornRU opened the pull request:
![image](https://user-images.githubusercontent.com/41479614/213918242-eb06b37b-a7c5-4914-9e4d-bf4c51b8ddcb.png)

## Описание
Новая цель - Сканирование Блюспейс Разломов
```
Научно-исследовательская станция расположена на месте дислокации блюспейс разлома.
Задача станции изучить все аномалии производимые разломом с сбором данных. 
Собранные данные отправятся на изучение центральным научно-исследовательским отделом Нанотрейзен.
```
![image](https://user-images.githubusercontent.com/41479614/210837943-859cbda9-0c2c-43f0-b596-80f1d27208a4.png)

Один из серверов должен полностью изучить разломы, чтобы цель считалась засчитанной.
В протолате и в принтере плат добавлены новые платы и T-ray сканнеры.

### Объекты:
#### Блюспейс разломы:
1. Блюспейс разлом - обычный разлом средней мощности. Умеренная скорость.
2. Блюспейс Разлом - разлом большой величины и мощности. Медленный
3. Блюспейс Туманность - крайне опасный огромный разлом. Крайне медленный.
4. Разлом-Близнец - появляется сразу в 2-х экземплярах. Аномалии исходящие от одного разлома, исходят и от другого. Небольшого размера.
5. Блюспейс Трещина - малый разлом в 4-х экземплярах. Быстрый. Наименее опасный в критических зонах.
![image](https://user-images.githubusercontent.com/41479614/210843628-a156c9dc-f927-49eb-81ab-2040745e8296.png)

Если не сканировать разломы - они безвредны. При сканировании начинается процесс создания событий и аномалий со случайными эффектами и непредсказуемыми результатами.

Разломы смещаются исключительно в безопасные места на станции и их невозможно остановить или замедлить.

В директивах с ЦК указывается какой разлом предстоит изучить и просканировать:
![image](https://user-images.githubusercontent.com/41479614/210843945-89357096-d38a-4593-8ea5-8acab090d7fb.png)


#### Сервер:
Спрайт от Элл Гууд
![image](https://user-images.githubusercontent.com/41479614/210842133-97565953-7e02-4cd1-bde1-9d2197d331da.png)
Хранилище данных. Собирает данные со всех сканнеров, позволяет контролировать количество очков собранных всеми сканерами.
При наличии нескольких сканнеров, очки исследования делятся между ними при получении.
Шкала результата - шкала выбирающая максимальное значение из всех серверов.
Количество серверов не влияет на число получаемых очков серверами. Но оно влияет на очки-стимуляции, равномерно деля их между всеми серверами.

При достижении заполненной шкалы - выдает награду 1 раз за одну цель.
При достижении заполненной шкалы стимуляции - позволяет "попытать удачу" и сыграть в рулетку с разломом, где с 50% шансом выпадет либо случайный предмет, либо аномалия. Выпадение происходит в самом блюспейс разломе.

ЕМП/Емаг переписывают протоколы безопасности сервера, из-за чего рулетка позволяет в 5 раз (зависит от делителя) меньше тратить очки на рулетку, но увеличивает шансы на появление отрицательного исхода рулетки.

##### Интерфейс:
![image](https://user-images.githubusercontent.com/41479614/210836642-965e77bc-f575-4344-9eda-6d1b8a6bd4ce.png)

Активация стимуляции, доступно 2 заряда:
![image](https://user-images.githubusercontent.com/41479614/211020902-5b637663-2fc3-430b-941e-499d2c34fb6c.png)

Статусы серверов:
![image](https://user-images.githubusercontent.com/41479614/210844360-b69637e7-b57b-47eb-a279-0bc0b151630d.png)

#### Сканнеры:
Спрайт от Элл Гууд
Виды сканеров:
1. Динамичный сканер - переносный и удобный сканер, который возможно свободно перемещать при откручивании и прикручивании пола. Малое время для устойчивости в критической зоне, малый радиус сканирования.
![image](https://user-images.githubusercontent.com/41479614/210842228-46302a01-f836-4c8f-83e9-a658f1bb1d55.png)

2. Статичный сканер - дорогой статичный сканер, требующий постоянного подключения к сети. Большой радиус сканирования и продолжительное время устойчивости в критической зоне.
Статичный сканер вращается при отсутствии разлома в зоне его досягаемости. И поворачивает свой сканер при нахождении разлома.
![image](https://user-images.githubusercontent.com/41479614/210842284-5abf4486-6d56-4e6f-ae71-06991bd16ff1.png)

Все сканнеры перед работой необходимо активировать рукой. 

При нахождении в критических зонах - случаются перегрузки, за время которых нужно успеть отключить сканнер. Если не успеть его отключить - произойдет критический ивент и он взорвется, из-за чего его придется повторно ремонтировать.

Сканнеры позволяют получать очки изучения. 
Чем дольше они работают и чем ближе находятся - тем больше очков и тем больше шансов вызвать случайное событие у разлома с другими аномалиями.
ЕМП/Емаг переписывают протоколы безопасности сканнеров, из-за чего критической зоной становится зона вне радиуса сканирования разлома, а бывшая критическая зона - становится допустимой территорией для сканирования.

#### T-Ray сканнеры:
Сканнеры с увеличенной длительностью импульса или расстояния. Изучаются вместе с открытием технологий в протолате.
1. Обычный
2. Увеличенная дальность
3. Увеличенная продолжительность импульса
4. Дальность и импульс.
5. Научный T-ray сканнер с еще большим увеличением. Открывается на 8-м уровне технологии магнетизма.
6. Экспериментальный T-ray сканнер. Хайриск. Находится у РД в шкафу. Не изучаемый и нельзя разобрать.
![image](https://user-images.githubusercontent.com/41479614/210842922-491f78e3-de71-4158-af6a-47b77882358e.png)

### Сборка:
#### Платы
![image](https://user-images.githubusercontent.com/41479614/211025023-aaf26e26-9001-432a-b997-8f26922c8882.png)
Печать находится в **Circuit Imprinter** в разделе **Research Machinery**
![image](https://user-images.githubusercontent.com/41479614/211024775-f9f7696d-acd0-4076-9bf3-beb55eabe251.png)

#### Требуемые компоненты:
1. **Сервер:**
10 metals, 5 glasss, 10 super capacitors, 2 phasic scanning modules and 20 cable coils.
2. **Портативный сканнер:**
5 metals, 2 super capacitors, 1 ultra-high-power micro-laser, 5 phasic scanning modules and 1 bluespace crystal.
3. **Статичный сканнер:**
30 metals, 8 super capacitors, 2 ultra-high-power micro-lasers, 10 phasic scanning modules and 4 bluespace crystals.

#### В протолате:
T-ray сканнеры в разделе **Equipment**

## Ссылка на предложение/Причина создания ПР
Авоут пройден. Дополнительная информация и детали в посте в Авоуте или АРАБской.
https://discord.com/channels/617003227182792704/755125334097133628/1048958642294100140

## Демонстрация изменений
**Работа разломов-близнецов, выпускают один и тот же эффект одновременно:**
![image](https://user-images.githubusercontent.com/41479614/210836906-897a5ea2-7b7e-4d77-9192-c955ece7e18d.png)

PhantornRU committed:
Инициализация BRS цели

PhantornRU committed:
Добавлены сканеры

Для печати + раундстартовый хайриск РД

PhantornRU committed:
Блюспейс разлом, движение разлома

PhantornRU committed:
Сканирование и нахождение в радиусе

PhantornRU committed:
Сервер, передача данных, критические значения

PhantornRU committed:
Ивенты и изучение, статичный сканер

PhantornRU committed:
Взаимодействия, шансы рандомизаций от максимумов

PhantornRU committed:
Поиск ближайшего разлома для сканера, фикс перекраски, взаимосвязи разломов-близнецов

PhantornRU committed:
Локальные ивенты, печать платы сканера, разломы для спавна

PhantornRU committed:
Убран локальный ивент перемешивания расы, доп. шанс на отказ от ивентов

PhantornRU committed:
Интерфейс сервера

PhantornRU committed:
Интерфейс, награды

PhantornRU committed:
Рулетка для траты полученных очков и получения дополнительных наград

PhantornRU committed:
Правки интерфейса при рассчетах

PhantornRU commented:
Завтра пофикшу перекрашивание разлома. А то чего-т он поломалься.
А также оторву гостам руки чтобы они больше не могли жмякать кнопочки на сервере.

Den109G commented:
Это что, теперь можно потрогать паралакс?

PhantornRU committed:
Цветовое изменение разломов

PhantornRU committed:
Фикс рулетки

Den109G commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -227,10 +227,11 @@
 	temp_list.Add(src)
-	message_admins("[name] Цвет: [color]")
+	var/temp_colour = rand_hex_color()
+	color = temp_colour
```
А просто `color = rand_hex_color()`?
temp_colour в этом проке просто два раза упоминается. Этот и `rift.color = temp_colour`. Может поменять на просто color?

PhantornRU replied:
имеет смысл, но ничего не поменяет по сути дела. Шо так шо так один и тот же цвет перекрашивается.

BeebBeebBoob replied:
@Den109G имел смысл если не юзалось в строке 248

PhantornRU committed:
Фиксы путей Т-рей сканнеров и платы сервера для сборки

BeebBeebBoob commented on code/__DEFINES/rift.dm:
```diff
@@ -0,0 +1,13 @@
+// Тип разлома
```
На английском

BeebBeebBoob commented on code/__DEFINES/rift.dm:
```diff
@@ -0,0 +1,13 @@
+#define CRACK_RIFT 5
+#define MAX_TYPES_RIFT 	5
+
+// Тип ивента
```
На английском

BeebBeebBoob commented on code/__DEFINES/sight.dm:
```diff
@@ -18,12 +18,15 @@
 #define INVISIBILITY_SPIRIT 50
 #define SEE_SPIRITS 50
 
+#define INVISIBILITY_ANOMALY 40	//значение видимости аномалии/разлома гостами и сканерами
```
На английском

BeebBeebBoob commented on code/game/gamemodes/steal_items.dm:
```diff
@@ -164,3 +164,8 @@
+/datum/theft_objective/experimantal_t_ray
+	name = "экспериментальный T-ray сканнер"
+	typepath = /obj/item/t_scanner/experimental
+	protected_jobs = list("Research Director")
```
Это не входило в сделку предожки

PhantornRU replied:
Убрал

BeebBeebBoob commented on code/game/objects/items/devices/scanners.dm:
```diff
@@ -22,15 +22,57 @@ REAGENT SCANNER
+	origin_tech = "magnets=8;engineering=5"
+	materials = list(MAT_METAL=500)
+
+/obj/item/t_scanner/experimental	//хайриск который нельзя разобрать, так как эта хуйня изобретена ну сами знаете кем.
```
На английском

PhantornRU replied:
зачем? В этом коде аглоязычники никогда рыться не будут. Это исключительно для нас фича.

BeebBeebBoob commented on code/game/objects/items/weapons/grenades/clusterbuster.dm:
```diff
@@ -199,7 +199,7 @@
 	payload = /obj/random/tool
 
-/obj/item/grenade/clusterbuster/tools
+/obj/item/grenade/clusterbuster/eng_tools
```
Смысл?

PhantornRU replied:
фикс мелкий.

BeebBeebBoob commented on code/modules/research/designs/equipment_designs.dm:
```diff
@@ -305,8 +305,48 @@
 	desc = "An advance use of a terahertz-ray to find any invisible biological creature nearby."
 	id = "sec_tray"
-	req_tech = list("combat" = 4, "magnets" = 7, "biotech" = 7)
+	req_tech = list("magnets" = 7, "biotech" = 7, "engineering" = 3)
```
Это входило в сделку?

PhantornRU replied:
Подтягивание под общие стандарты технологий и T-ray сканнера. Боевые технологии вообще никак не связаны с обнаружением. Биотеч - да. Комбат - нет.

BeebBeebBoob commented on code/modules/station_goals/brs.dm:
```diff
@@ -0,0 +1,85 @@
+GLOBAL_LIST_INIT(bluespace_rifts_server_list, list())
+
+// Сканер Блюспейс Разлома
+// Цель для исследования аномального блюспейс разлома с созданием портативных и статичных сканеров
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+//==========================================
+//============ Локальные ивенты ============
+//==========================================
+//Выбор локального ивента
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+		if(2)
+			local_random_grenade(objects)
+
+//Выбор локального связанного ивента
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+
+//============ Телепорты ============
+
+//Телепорты в небольшом радиусе
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+			var/turf/T = get_turf(H)
+			do_teleport(H, get_turf(temp_object))
+			do_teleport(temp_object, T)
+			investigate_log("teleported reshuffle [key_name_log(H)] and [key_name_log(temp_object)]", INVESTIGATE_TELEPORTATION)
```
```suggestion
			investigate_log("switched places [key_name_log(H)][COORD(H)] and [key_name_log(temp_object)][COORD(temp_object)]", INVESTIGATE_TELEPORTATION)
```
И не объект, а моб

PhantornRU replied:
Поправил

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+//============ АОЕ эффекты ============
+/obj/brs_rift/proc/local_explosive(var/list/objects)
+	for(var/obj/O in objects)
+		var/fs = force_sized
+		explosion(O.loc, 0, 1, fs, 2*fs, flame_range = 3*fs, cause = O)	//Вставлять O.loc???
```
Смысл переменной? 
Суй сразу туда его, убирай
А O.loc на O

```suggestion
		explosion(O, 0, 1, force_sized, 2*force_sized, flame_range = 3*force_sized, cause = O)
```

PhantornRU replied:
Если совать просто O, он в некоторых моментах не может получить TURF и просто ломается, не воспроизводя функцию. Я ж изначально так и делал

PhantornRU replied:
```suggestion
		var/fs = force_sized
		explosion(O.loc, 0, 1, fs, 2*fs, flame_range = 3*fs, cause = O)
```

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+
+/obj/brs_rift/proc/local_emp(var/list/objects)
+	for(var/obj/O in objects)
+		var/fs = force_sized
+		empulse(O.loc, fs, 2*fs, TRUE, name)
```
```suggestion
		empulse(O, force_sized, 2*force_sized, TRUE, name)
```

PhantornRU replied:
Для более понятного кода - лучше оставить fs чтобы строчки не было на пол экрана. Это не в одном месте так.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+	for(var/obj/O in objects)
+		make_random_grenade_prime(choosen, O)
+
+//Выбираем гранату и сразу же её подрываем, тем самым "воруя" её эффект хе-хе-хе
```
Молодец. А потом будешь объяснять людям почему аномалия кидается гранатами.
Не говоря что ты весь код разломал на несколько функции

BeebBeebBoob replied:
Хотя я хз. Там же задержка перед взрывом. Но это самый ужасный костыль.

PhantornRU replied:
Там нет задержки перед взрывом. Там мгновенный активатор

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+
+
+//============ Изменение облика ============
+//рандомизация днк
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,389 @@
+		if(istype(H) && H.stat != DEAD && !H.notransform)
+			randomize_species(H)
+
+//Рандомизация расы и её преференса
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+	new /obj/effect/portal(temp_loc, null, null, 40)
+
+
+//================ Объекты ================
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+	info = "<b>Долгожданные научные исследования блюспейс разлома, продвигающие науку изучения Синего Космоса далеко вперед. \nВ записке написана тарабарщина на машинном языке. \nТребуется деструктивный анализ.</b>"
+	origin_tech = "bluespace=9;magnets=8"
+
+//блюспейс толкан
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+	icon_state = "bluespace_toilet00"
+	var/teleport_sound = 'sound/magic/lightning_chargeup.ogg'
+
+/obj/structure/toilet/bluespace/brs
+	name = "Воронка Бездны Синего Космоса"
+	desc = "То, ради чего наука и была создана и первый гуманоид ударил палку о камень. Главное не смотреть в бездну."
+	icon_state = "bluespace_toilet00-NT"
```
А вот честно. Нету что либо получше чем сортирный юмор?
Пупс, давай как адекват, придумай нормальную награду. К примеру универсальный телепортер или возможность как-то стабилизировать аномалии. Чтобы интересно и без юмора на уровне туалетного

PhantornRU replied:
В том-то и дело, что это цель для всей станции. И кроме как технологий всей станции ничего и не нужно. Иначе это поспособствует только игре одного человека, забравшим себе условный телепортер.
Плюс дополнительная награда в виде технологий открывает возможность печати научных T-рей сканнеров и высокобюджетных блюспейс технологий (которые отсутствуют, но которые якобы нельзя в нормальной игре получить). И буквально, из-за педальского отношения в отказе выдавать игрокам "что-то ПГшное", в ходе совещания, пришел к этому абсурду.

Стабилизация разлома тоже не имеет смысла, т.к. достаточно её не сканировать, чтобы она не спавнила аномалии.

Поэтому только такое. Максимально не нужное и шуточное.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+
+		if(do_after(user, 100, target = src))
+			playsound(loc, teleport_sound, 100, 1)
+			teleport(1)
+
+/obj/structure/toilet/bluespace/proc/teleport(var/range_dist = 1)
```
Мог просто teleport() без единицы

PhantornRU replied:
не мог, оно не только там используется, при разрушении там дистанция вроде под 9.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+		/obj/random/therapy = 25,
+		/obj/random/carp_plushie = 25,
+
+		//Игрушки которые ты скорее всего никогда не получишь и ненужная мелочевка
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+		/obj/item/toy/syndicateballoon = 5,
+		/obj/item/stack/telecrystal = 1,
+
+		//Игрушки
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+	name = "brs loot"
+	lootcount = 1
+	loot = list(
+		//Тип предмета, его вес
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+		/obj/structure/toilet/bluespace,
+
+
+		//Не самые ценные вещи по прежнему могут выпасть, хоть и с небольшим шансом суммы их всех:
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+		/obj/item/clothing/head/collectable/xenom,
+		/obj/item/clothing/head/collectable/petehat,
+
+		//Магические вещи блюспейса с наиминимальнейшим шансом для появления
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_reward.dm:
```diff
@@ -0,0 +1,234 @@
+		/obj/item/spellbook/oneuse/mime/fingergun/fake,
+		/obj/structure/closet/crate/necropolis/tendril,
+
+		//Бабки, бабки, сука, бабки!
```
На английском

PhantornRU committed:
правки и решение конфликта

Решение конфликта
Update code/modules/station_goals/brs/brs_event_container.dm
Update code/modules/station_goals/brs/brs_server.dm

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+//Блюспейс разлом для создания веселья и беготни на станции
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	layer = MASSIVE_OBJ_LAYER
+	alpha = 180
+
+	//для отображения T-ray сканнера
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	invisibility = INVISIBILITY_ANOMALY
+	level = 1
+
+	var/timespan = 20 MINUTES	// промужуток времени смены направления
+	var/required_time_per_tile = 0	// необходимо времени перед движением на тайл
+	var/counter_direction_time		// счетчик времени перед сменой направления
+	var/counter_move_time 			// счетчик времени перед движением
+
+	var/force_sized = 5		// размер разлома и критической зоны
+	var/dir_move = 0		// направление
+	var/dir_loc = null		// место направления
+
+	var/type_rift = DEFAULT_RIFT
+	var/num_related_rifts = 1	//Сколько разломов может быть связано
+	var/related_rifts_list = list()	// связанные разломы (пр. разломы-близнецы)
+	var/anomaly_mod = 1.5		//Модификатор для контейнера ивентов
+	var/event_chance = 100	//Шанс появления ивентов для избегания флуда ивентами
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	required_time_per_tile = round(timespan/dist)
+
+/obj/brs_rift/attackby(obj/item/I, mob/living/user, params)
+	to_chat(user, "<span class='danger'>Невозможно взаимодействовать с разломом!</span>")
```
Ломом по разлому
Разлом:

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	to_chat(user, "<span class='danger'>Невозможно взаимодействовать с разломом!</span>")
+	return FALSE
+
+//Работа с весами в контейнерах
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+//Сканеры для исследования разломов и передачи данных на сервера
+//=============================
+//Портативный сканер 1х1
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+	var/obj/brs_rift/rift_for_scan = null	//концентрация на разломе для сканирования
+	var/rift_range = 0						//макс. расстояния на выбранном разломе
+
+	var/static/gid = 0
```
?????

PhantornRU replied:
Статичный тип данных.
Если вкратце, общая переменная для всех объектов, позволяющая нормально выдавать уникальный ID в моем проке.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+		if(prob(event_chance)) //* rift_range/dist)))
+			rift_for_scan.event_process(FALSE, dist, rift_range)
+
+	//деление очков на все сервера
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+			S.research_process(temp_points)
+
+/obj/machinery/brs_scanner/proc/critical_process(var/dist)
+	//Восстановление критического порога
+	if (counter_critical_time + critical_time * 2 < world.time)
+		counter_critical_time = 0
+
+	//Начало отсчета
+	if (counter_critical_time == 0)
+		counter_critical_time = world.time + critical_time
+
+	//Прохождение критического порога
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+		icon_state = "[prefix]-broken"
+		return
+
+	if (anchored)
+		if (toggle)
+			if (active)
+				if (emagged)
+					icon_state = "[prefix]-act-emagged"
+				else
+					icon_state = "[prefix]-act"
+			else
+				icon_state = "[prefix]-on"
+		else
+			icon_state = "[prefix]-anchored"
+	else
+		icon_state = prefix
```
Мои глаза. Убери лестницу.

PhantornRU replied:
как именно? Елочка вполне неплохо смотрится.

Свитч здесь не вариант, а каждый раз проверять или return'ать - не самое лучшее решение и только увеличит количество строк.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+	else
+		icon_state = prefix
+
+//Взаимодействия, разбор
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+	stat &= ~BROKEN
+	obj_integrity = max_integrity
+
+//Выдвижение и задвижение д.сканнера, активация ст.сканнера
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+		update_icon()
+	return TRUE
+
+//Перезапись протоколов безопасности.
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+/obj/machinery/brs_scanner/emag_act(mob/user)
+	if(!emagged)
+		rewrite_protocol()
+		to_chat(user, "<span class='warning'>@?%!№@Протоколы безопасности сканнера перезаписаны@?%!№@</span>")
```
-_-

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+		rewrite_protocol()
+		to_chat(user, "<span class='warning'>@?%!№@Протоколы безопасности сканнера перезаписаны@?%!№@</span>")
+
+/obj/machinery/brs_scanner/emp_act(severity)
+	if(!emagged && prob(40 / severity))
+		rewrite_protocol()
```
Емаг лох

BeebBeebBoob replied:
И сканнер тоже лох

PhantornRU replied:
Ибо нехуй держать сканнер под разломом, который пускает ЕМП ивент из-за другого сканнера неподалеку :))

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+		return TRUE
+	return FALSE
+
+// Составные компоненты
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+	RefreshParts()
+
+//=============================
+//Статичный сканер 3х3
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+	anchored = TRUE
+	density = TRUE
+	max_integrity = 500
+	critical_time = 30 SECONDS	//время до поломки в критических условиях
+	max_range = 50	//максимальное расстояния для исследований
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+		return
+	to_chat(user, "<span class='notice'>[name] разобран на составные компоненты.</span>")
+
+// Составные компоненты
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+//Сервер для контроля спутников
```
На английском

BeebBeebBoob replied:
Стоп спутников???

PhantornRU replied:
```suggestion
//Сервер для контроля сканнеров и наблюдения за выполнением цели
```

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+		return
+	icon_state = active ? "[prefix]-act" : "[prefix]-on"
+
+//==========Взаимодействия========
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+	stat &= ~BROKEN
+	obj_integrity = max_integrity
+
+// Составные компоненты
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+	component_parts += new /obj/item/stock_parts/capacitor/super(null)
+	RefreshParts()
+
+//Перезапись протоколов безопасности.
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+		rewrite_protocol()
+		to_chat(user, "<span class='warning'>@?%!№@Протоколы безопасности сканнера перезаписаны@?%!№@</span>")
+
+/obj/machinery/brs_server/emp_act(severity)
+	if(!emagged && prob(40 / severity))
+		rewrite_protocol()
```
Емаг и защита от Эми лохи

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+
+	ui_interact(user)
+
+//Интерфейс
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_server.dm:
```diff
@@ -0,0 +1,283 @@
+				visible_message("<span class='notice'>Исследование завершено.</span>")
+				. = TRUE
+			else
+				playsound(loc, 'sound/machines/buzz-two.ogg', 100, 1)
```
На английском

BeebBeebBoob commented:
Я пока что на телефоне просмотрел.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+			timespan = 30 MINUTES
+			name = "Блюспейс Туманность"
+
+	transform = matrix(force_sized, 0, 0, 0, force_sized, 0) //+ перекрас?
```
А может без transform?
Сделаешь нормальные спрайты, ибо смотря на тех размытых слаймов... кошмар

BeebBeebBoob replied:
Это также будет понятно людям и приятнее.

PhantornRU replied:
на спрайты спрайтер нужен. А у нас лапки. Вдовесок они полупрозрачные и не видимы невооруженным глазом. Размытость в итоге только в плюс с сливанием при прозрачности играет.

PhantornRU replied:
![image](https://user-images.githubusercontent.com/41479614/211131865-591c157a-1947-491f-90f2-19f7c961693d.png)

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	var/related_rifts_list = list()	// связанные разломы (пр. разломы-близнецы)
+	var/anomaly_mod = 1.5		//Модификатор для контейнера ивентов
+	var/event_chance = 100	//Шанс появления ивентов для избегания флуда ивентами
+
```
```suggestion
//For admins
```
Иначе по другому не объяснит смысл существование данных подтипов

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	force_sized = 9
+	timespan = 30 MINUTES
+
+/obj/brs_rift/twin/t_static	//Для тестов и баловства
```
На английском

BeebBeebBoob commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+				else
+					M.weight_mod /= mod
+
+/obj/brs_rift/proc/make_related_list(var/n = 1, var/type_rift)
```
Что за n?

PhantornRU replied:
```suggestion
/obj/brs_rift/proc/make_related_list(var/num = 1, var/type_rift)
```

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+
+/obj/machinery/brs_scanner/proc/find_nearest_rift()
+	var/obj/brs_rift/min_rift
+	var/min_dist = max_range*2
```
а не max_dist?
смотря по строке 217 и 218
```
var/temp_range = max_range + rift.force_sized
if (dist <= temp_range && dist <= min_dist)
```
Да при этом, в чём смысл если мы и так проверяем до max_range плюс размер rift. максимальный rift 9 что меньше сканера на 10
/shrug
достаточно проверка что rift в пределах сканера плюс его размера.

Пример:
Сканер = 10
Рифт = 9
Дистанция = 14 для примера
if( 14 <= 19 и 14 <= 10*2) 
/shrug

PhantornRU replied:
И по итогу сломаем логику и добавим баг, т.к. нам не нужно max_range каждый раз умножать на 2, а только 1 самый первый раз. В дальнейшем он приобретает дистанцию.
```
var/min_dist = max_range*2
for (бла-блабла)
	if (dist <= temp_range && dist <= min_dist)
		min_dist = dist
```

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+					/obj/item/stack/ore/bluespace_crystal = 4
+					)
+
+/obj/machinery/brs_scanner/s_static
```
```suggestion
/obj/machinery/brs_scanner/static
```
brs scanner scanner static

PhantornRU replied:
static - тип данных.
Не должно быть путаницы.

BeebBeebBoob commented on code/modules/station_goals/brs/brs_scanner.dm:
```diff
@@ -0,0 +1,366 @@
+				critical_process(dist)
+			else
+				scanner_process(dist)
+		else	//При ЕМАГе, возможно сканирование в любом радиусе. Вне его - критическая зона
```
На английском

PhantornRU commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+	var/temp_colour = rand_hex_color()
+	color = temp_colour
+
+	if(n <= 1)
```
```suggestion
	if(num <= 1)
```

PhantornRU commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+
+	for(var/obj/brs_rift/rift in GLOB.bluespace_rifts_list)
+		if (rift.type_rift == type_rift && rift != src)
+			if (length(rift.related_rifts_list) >= n)
```
```suggestion
			if (length(rift.related_rifts_list) >= num)
```

PhantornRU commented on code/modules/station_goals/brs/brs_rift.dm:
```diff
@@ -0,0 +1,248 @@
+			if (length(rift.related_rifts_list) >= n)
+				continue
+			temp_list.Add(rift)
+			if(length(temp_list) >= n)
```
```suggestion
			if(length(temp_list) >= num)
```

PhantornRU commented:
@BeebBeebBoob к чему делать комментарии на английском на русскоязычном проекте с русскоязычными разработчиками, в код которого полезут именно русскоязычные кодеры?
У нас нету англиков. А заставлять переводить русскоязычные комментарии на английский для наших же русскоязычного проекта, откуда фичи никуда не утекут и никогда не будут использоваться англиками - определенно странно.
Делаем мы всё же для себя. Иначе вернемся в эру переводов с наполеоновскими планами мультиязычности, где мультиязычность ну абсолютно никому не нужна.

А кодировка буквально не успеет испортиться к моменту всяких переездов с поломкой символов, т.к. потенциально мы всеравно будем двигаться на тот же СС14 через годик с лишним.

Уж кого, но я против перевода комментариев на английский. Это полностью наша разработанная фича и в неё будут рыться наши разрабы и сам я. И другие разрабы, и я, не должен/не должны сидеть и думать какую хуйню я там напереводил чтобы повторно напереводить обратно, чтобы понять что куда идет.

С названиями объектов, переменных - я согласен. Но они и так на английском. А комментарии на то и комментарии. Они исходят от того кто их пишет. И я лично не говорю в повседневности на английском. Мне это было бы крайне неудобно. Особенно если в будущем я вдруг соберусь расширять эту цель.

dikers93 commented:
> @BeebBeebBoob к чему делать комментарии на английском на русскоязычном проекте с русскоязычными разработчиками, в код которого полезут именно русскоязычные кодеры? У нас нету англиков. А заставлять переводить русскоязычные комментарии на английский для наших же русскоязычного проекта, откуда фичи никуда не утекут и никогда не будут использоваться англиками - определенно странно. Делаем мы всё же для себя. Иначе вернемся в эру переводов с наполеоновскими планами мультиязычности, где мультиязычность ну абсолютно никому не нужна.
> 
> А кодировка буквально не успеет испортиться к моменту всяких переездов с поломкой символов, т.к. потенциально мы всеравно будем двигаться на тот же СС14 через годик с лишним.
> 
> Уж кого, но я против перевода комментариев на английский. Это полностью наша разработанная фича и в неё будут рыться наши разрабы и сам я. И другие разрабы, и я, не должен/не должны сидеть и думать какую хуйню я там напереводил чтобы повторно напереводить обратно, чтобы понять что куда идет.
> 
> С названиями объектов, переменных - я согласен. Но они и так на английском. А комментарии на то и комментарии. Они исходят от того кто их пишет. И я лично не говорю в повседневности на английском. Мне это было бы крайне неудобно. Особенно если в будущем я вдруг соберусь расширять эту цель.

Это все хочет не бип, а димач с Азизом, попросту не пропускают руские такие подсказки

PhantornRU committed:
Рефактор, добавлен охотничий разлом

PhantornRU committed:
Экспериментальный T-ray убран из хайрисков для кражи.

PhantornRU committed:
Чистка комментариев и перевод на проклятый английский

NOOOOOOOOOOOOO

PhantornRU commented:
Я сломался и всё-таки сделал переводы комментариев :(

PhantornRU committed:
Правка координат при смене от Бибубубубубупапабибупа

PhantornRU committed:
Емаг сканнера уменьшает очки для рулетки и шанс получения положительного приза, новая иконка емаг сканнера

Dimach commented on code/modules/research/designs/equipment_designs.dm:
```diff
@@ -305,8 +305,48 @@
 	category = list("Equipment")
+
+/datum/design/tray_scanner_range
+	name = "Расширенный Т-сканнер"
```
Так, это конечно все волшебно но тебя ученые просто возненавидят за то что при поиске в менюхе протолата надо будет переключать раскладку, сделай хотя бы в нём на английском все

PhantornRU replied:
Ок, названия на английский переделаю

Dimach commented on code/modules/research/designs/machine_designs.dm:
```diff
@@ -551,3 +551,33 @@
 	category = list ("Misc. Machinery")
+
+/datum/design/brs_server
+	name = "Machine Design (Сервер сканирирования разлома)"
```
Тоже самое что выше

Dimach commented on code/modules/station_goals/brs/brs_event_container.dm:
```diff
@@ -0,0 +1,290 @@
+	var/obj/item/grenade/gr
+	switch(choosen)
+		if(1)
+			gr = new /obj/item/grenade/smokebomb(new_loc)
```
А почему бы эти тайппасы просто в list не сложить, и обращаться к ним по индексу?

PhantornRU replied:
А ведь действительно!  Так и сделаю.

PhantornRU replied:
Вынес в отдельный список

PhantornRU committed:
Чертежи протолата для сканеров переведены на английский

PhantornRU committed:
Перевод на английский чертежей сканеров и сервера

PhantornRU committed:
Вынесение гранатных эффектов в отдельный список

PhantornRU committed:
немного повышаем шанс до 25% при емаге

PhantornRU committed:
Багфиксы, ротация (подробнее в описании)

фикс интерфейса,
добавлена "передышка" от ивентов, прикручивание с.сканера,
снижение шанса спавна аномалии,
убрана цель из ротации на время тестов

PhantornRU committed:
Решение конфликта ТГУИ

PhantornRU committed:
Передышка дается после спавна глобал-ивента

PhantornRU committed:
Правка ошибки логики

PhantornRU committed:
Увеличение получаемых очков при сканировании

PhantornRU committed:
Стимуляция дает дополнительные очки исследования

PhantornRU commented:
@Dimach залил пак фиксов-времянок. Убрал из ротации, оставил для щитспавнов. Лучше побыстрее закинуть, чтобы оно в ротации не выпадало. То что щас на Тестмердже слишком жесткое для людей оказалось. Поэтому пускай пока под присмотрами педалей теперь только ставится. После того как окончу тестирования и завершения цельки уже буду выкатывать. Мож через полмесяца-месяц, как время выйдет, ибо до 10-го у меня щас завал будет небольшой.

PhantornRU committed:
Теперь очки нормально распределяются по серверам, но очки для рулеток равномерно делятся

PhantornRU commented:
Вкратце о цели:
![image](https://user-images.githubusercontent.com/41479614/213918224-2b0d2aa3-aaed-4a85-bd12-855319dc97e5.png)

PhantornRU committed:
Багфиксы и ребаланс: Возвращение в пул для выпадения, корректировка рулетки, дистанции критического сканнера, фиксы ивент панели

PhantornRU committed:
Мелкофикс на починку

PhantornRU commented:
ПР закрыт, так как я больше не собираюсь кодить для Парадайза.

PhantornRU commented:
Жалко конечно что на этот ПР потратил кучу времени и новогодние праздники. Но если кто-то захочет продолжить - я буду не против. Но никто не захочет. Кодить для парадайза уж явно не стоит для сохранения своих нервов и достоинства.
